File: langCompare/pas/convert.pas
program convert;

const
	digits : string = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';

function find(c : char; str : string) : integer;
var
	i : integer;
begin
	for i := 1 to length(str) do 
	begin
		if c = str[i] then exit(i-1);
	end;
	find := 255;
end;

function reverse(inStr : string) : string;
var
	out : string;
	c   : char;
begin
	out := '';	
	for c in inStr do
	begin
		out := c + out;
	end;	
	reverse := out;
end;

function strToInt(str : string; base : integer) : integer;
var
	number : integer = 0;
	multi  : integer = 1;
	revStr : string;
	c      : char;
	digit  : integer;
begin
	revStr := reverse(str);
	for c in revStr do
	begin
		digit := find(c, digits);
		number := number + digit * multi;
		multi := multi * base;	
	end;
	strToInt := number;
end;

function intToStr(number : integer; outBase : integer ) : string;
var
	digit  : integer;
	outStr : string;
begin
	if number = 0 then exit('0');	
	
	outStr := '';
	while number > 0 do
	begin
		digit  := number mod outBase;
		outStr := digits[digit+1] + outStr;
		number := number div outBase;
	end;
	
	intToStr := outStr;
end;

function convert(inStr : string; inBase : integer; outBase : integer) : string;
begin
	convert := intToStr(strToInt(inStr,inBase),outBase);
end;

begin
	write(paramStr(1));
	write(' -> ');
	writeln(convert(paramStr(1), strToInt(paramStr(2), 10), strToInt(paramStr(3), 10)));
end.


{
procedure reverseInPlace(var str : string);
var
	i : integer;
	t : char;
begin
	for i := 1 to length(str) div 2 do 
	begin
		t := str[i];
		str[i] := str[length(str)-(i-1)];
		str[length(str)-(i-1)] := t;
	end;
end;	
}

File: kotlin/Convert.kt
// lnc exercise in Kotlin with some comments
// For pedagogical reasons, this code is kept close to the Java implementation to make
// a comparison possible.
// val declarations are immutable values, var declarations are mutable, non-final variables
// const is used for compile-time constants
const val digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

fun reverse(str: String): String {
    var revStr = "" // String type is inferred from the initializer
    for (c in str.toCharArray()) {
        revStr = c + revStr
    }
    return revStr
}

fun find(c: Char, str: String): Int {
    for (index in str.indices) {
        if (c == str[index]) return index
    }
    return 255
}

fun strToInt(inStr: String, inBase: Int): Long {
    var number: Long = 0 // We need the type spec since 0 would be Int otherwise
    var multi = 1L // But this is also possible. Multi is a Long now
    for (element in reverse(inStr)) { // Compare this to the loop in find(..)
        val digit = find(element, digits)
        number += digit * multi
        multi *= inBase
    }
    return number
}

fun intToStr(number: Long, outBase: Int): String {
    var n = number
    var revOutStr = ""
    if (n == 0L) return "0"
    while (n > 0) {
        val digit = (n % outBase).toInt()
        revOutStr += digits[digit]
        n /= outBase
    }
    return reverse(revOutStr)
}

fun convert(inStr: String, inBase: Int, outBase: Int): String {
    return intToStr(strToInt(inStr, inBase), outBase)
}

fun main(args: Array<String>) {
    println(args[0] + " -> " + convert(args[0], strToInt(args[1], 10).toInt(), strToInt(args[2], 10).toInt()))
}
